Custom domains documentation (#739)
fermyon/developer: content/cloud/custom-domain.md

Reviewer: Just a note: this page has a date of publish set to a time in the future. If this PR is merged before `date = "2023-07-26T12:00:00Z"` UTC then the menu in the sidebar that points to this page will be a 404 error.

Please consider the timing around publishing this content/page vs the whole PR and make any adjustments in accordance with official announcements/press releases and so forth.

If you do not own a domain (or are **not** interested in purchasing one) but would like to customize your Spin application's domain name while still keeping the `fermyon.app` apex domain, then Custom Fermyon Subdomains may be a good option for you. Check out the [Custom Fermyon Subdomains tutorial](./custom-fermyon-subdomain.md) for more instructions. 

If you do not own a domain and you **are** interested in purchasing one, you can purchase one through a domain registrar like [Namecheap](https://www.namecheap.com/) or [Google Domains](https://domains.google). At this time, Fermyon Cloud is not a domain registrar meaning you'll have to use an external registrar.

>> If you are unsure what some of the terms on this page mean, please see the section called [Terminology](#terminology) at the bottom of this page.

> To take advantage of Fermyon Cloud Custom Domain, you must delegate your domain to Fermyon Cloud. For a full set of instructions, please follow the [Ferymon Cloud Custom Domains tutorial](./custom-domains-tutorial.md). 

Reviewer 2: I think this merits its own subhead, because it's the only how-to information currently available, and it's easy to miss (well, at least, it was easy for me!):

```suggestion
## Setting Up Custom Domains

To take advantage of Fermyon Cloud Custom Domains, you must delegate your domain to Fermyon Cloud. For a full set of instructions, please follow the [Ferymon Cloud Custom Domains tutorial](./custom-domains-tutorial.md). 
```

Author: Love making it super clear!

title = "Custom Domains"
template = "cloud_main"

> date = "2023-07-26T12:00:00Z"

Reviewer: ```suggestion
date = "2023-07-26T16:00:00Z"
```

  - [Next Steps](#next-steps)
  - [Terminology](#terminology)

Every Fermyon Cloud application has a default domain name in the format `<app-name>-<random-string>.fermyon.app`. The Fermyon Cloud Custom Domains feature provides a way for you to use your own domain names; providing your users with human-friendly access to your Fermyon Cloud applications. Another advantage of using Fermyon Cloud Custom Domains is that your Spin application can have a stronger sense of branding and identity. For example, for [Finicky Whiskers: The World's Most Adorable Manual Load Generator](https://www.fermyon.com/blog/finicky-whiskers-part-1-intro) we have moved away from using the default domain name `finicky-whiskers-6cavavfh.fermyon.app`, and we much prefer using the custom domain `finickywhiskers.com`. 

If you do not own a domain (and are **not** interested in purchasing one) but would like to customize your Spin application's domain name within the `fermyon.app` apex domain, then Custom Fermyon Subdomains may be a good option for you. Check out the [Custom Fermyon Subdomains tutorial](./custom-fermyon-subdomain.md) for more instructions. 

> If you do not own a domain and you **are** interested in purchasing one, you can purchase one through a domain registrar like [Namecheap](https://www.namecheap.com/) or [Google Domains](https://domains.google). Fermyon Cloud is not a domain registrar meaning you'll have to use an external registrar.

Reviewer 3: Google Domains is EOL.

Reviewer: Thanks @radu-matei , great catch!

> If you do not own a domain and you **are** interested in purchasing one, you can purchase one through a domain registrar like [Namecheap](https://www.namecheap.com/) or [Google Domains](https://domains.google). Fermyon Cloud is not a domain registrar meaning you'll have to use an external registrar.

Reviewer: ```suggestion
If you do not own a domain and you **are** interested in purchasing one, you can purchase one through a domain registrar like [Namecheap](https://www.namecheap.com/) or [GoDaddy](https://www.godaddy.com/). Fermyon Cloud is not a domain registrar meaning you'll have to use an external registrar.
```

> If you do not own a domain and you **are** interested in purchasing one, you can purchase one through a domain registrar like [Namecheap](https://www.namecheap.com/) or [Google Domains](https://domains.google). At this time, Fermyon Cloud is not a domain registrar meaning you'll have to use an external registrar.

Reviewer 4: > At this time,

Just checking to make sure: do we want to lead the reader onto the fact that we are considering this? If so, is there something we want to provide the user to ask for feedback?

Reviewer: Good call Matt, if we do become a registrar we will just say we are ... then.

> If you do not own a domain and you **are** interested in purchasing one, you can purchase one through a domain registrar like [Namecheap](https://www.namecheap.com/) or [Google Domains](https://domains.google). At this time, Fermyon Cloud is not a domain registrar meaning you'll have to use an external registrar.

Reviewer: ```suggestion
If you do not own a domain and you **are** interested in purchasing one, you can purchase one through a domain registrar like [Namecheap](https://www.namecheap.com/) or [Google Domains](https://domains.google). Fermyon Cloud is not a domain registrar meaning you'll have to use an external registrar.
```

### Next Steps

* Follow the [Ferymon Cloud Custom Domains tutorial](./custom-domains-tutorial.md) to add your custom domain.
* Visit [Technical FAQ](./faq.md) for troubleshooting assistance.
* Check out [Pricing](./pricing-and-billing.md) for billing-related questions.

### Terminology

* Apex Domain: the `finickywhiskers.com` part in `www.finickywhiskers.com`.
* Custom Domain: a personalized and unique address that individuals, businesses, or organizations can register and use for their workloads. In the context of Fermyon Cloud, this is the domain name you provide. 
* Custom Fermyon Subdomain: a customized name you select following the pattern  `<custom-name>.fermyon.app`. 
* Domain Name: human-readable address to identify a workload on the internet. For example, `finicky-whiskers-6cavavfh.fermyon.app` or `www.finickywhiskers.com`.
* Domain Name Service (DNS) provider: a service that manages and maintains a database of domain names and their corresponding IP addresses. 

> * Domain Registrar: a company or organization accredited by domain name registries to allow individuals and businesses to register and manage domain names for their websites. For example, [Namecheap](https://www.namecheap.com/) or [Google Domains](https://domains.google/) are domain registrars. 

Reviewer: ```suggestion
* Domain Registrar: a company or organization accredited by domain name registries to allow individuals and businesses to register and manage domain names for their websites. For example, [Namecheap](https://www.namecheap.com/) or [GoDaddy](https://www.godaddy.com/) are domain registrars. 
```

- [Terminology](#terminology)
- [Relationship Between Custom Domains And Spin Applications](#relationships-between-custom-domains-and-spin-applications)
- [Next Steps](#next-steps)

> ## Fermyon Cloud Custom Domains Explained

Reviewer: Thanks @itowlson
Agreed, this heading will look like a duplicate so I am removing in the next push to this branch.

> Fermyon Cloud Custom Domains empowers developers to bring their own domain name to Fermyon Cloud. A domain name translates to a unique IP addresses, serving as a "human-friendly" entrypoint for websites and other public facing workloads. Every Spin application is assigned a default domain name with the format `<app-name>-<random-string>.fermyon.app`. However, developers may want to use a custom domain name with their Spin application to improve discoverability and provide a stronger sense of branding and identity. For example, rather than using the default domain name `finicky-whiskers-6cavavfh.fermyon.app`, a developer may prefer using their custom domain `finickywhiskers.com` to direct traffic to their Spin application. 

Reviewer: Thanks, I have updated and removed empowers.

> Fermyon Cloud Custom Domains empowers developers to bring their own domain name to Fermyon Cloud. A domain name translates to a unique IP addresses, serving as a "human-friendly" entrypoint for websites and other public facing workloads. Every Spin application is assigned a default domain name with the format `<app-name>-<random-string>.fermyon.app`. However, developers may want to use a custom domain name with their Spin application to improve discoverability and provide a stronger sense of branding and identity. For example, rather than using the default domain name `finicky-whiskers-6cavavfh.fermyon.app`, a developer may prefer using their custom domain `finickywhiskers.com` to direct traffic to their Spin application. 

Reviewer: Thanks, I have reworded this with a mix of ideas from here.

> > If you don't already have a custom domain, you can purchase one through a domain registrar, like [Namecheap](https://www.namecheap.com/) or [Google Domains](https://domains.google/?gad=1&gclid=Cj0KCQjwzdOlBhCNARIsAPMwjbyNg3a3GcrQjbs3vBSDg4AWnnEkc96mJupyg8nqqiKyxVCyaePRXfAaAjRLEALw_wcB&gclsrc=aw.ds). At this time, Fermyon Cloud is not a domain registrar meaning you'll have to use an external registrar. 

Reviewer: Agreed, I have removed the query string `/?gad=1&gclid=Cj0KCQjwzdOlBhCNARIsAPMwjbyNg3a3GcrQjbs3vBSDg4AWnnEkc96mJupyg8nqqiKyxVCyaePRXfAaAjRLEALw_wcB&gclsrc=aw.ds`
We can put it back if we meant for it to be there in the first place but as I van points out we may want to disclose any purchase affiliation promotion tracking etc.

To take advantage of Fermyon Cloud Custom Domain, you must delegate your domain to Fermyon Cloud. For a full set of instructions, please follow the [Ferymon Cloud Custom Domains tutorial](./custom-domains-tutorial.md). 

> If you do not own a domain (or are not interested in purchasing one), but would like to customize your Spin application's domain name while still keeping the `fermyon.app` apex domain, then Custom Fermyon Subdomains may be a good option for you. Check out the [Custom Fermyon Subdomains tutorial](./custom-fermyon-subdomain.md) for more instructions. 

Reviewer: Thanks, I have brought this up and laid out the options fairly clearly; using almost identical starts to the sentences of the two involved paragraphs.

### Terminology

> * Apex Domain: the `finickwhiskerse.com` part in `www.finickywhiskers.com`

Reviewer: Thanks Ivan, I have updated this in the next push (with the other changes).

* Custom Domain: a personalized and unique address that individuals, businesses, or organizations can register and use for their workloads. In the context of Fermyon Cloud, this is the domain name you provide. 

> * Custom Fermyon Subdomain: a customized name you select following the pattern  `<custom-name>.fermyon.app`. 

Reviewer: I have moved this to a reference section to negate any confusion.

* Domain Name: human-readable address to identify a workload on the internet. For example, `finicky-whiskers-6cavavfh.fermyon.app` or `www.finickywhiskers.com`
* Domain Name Service (DNS) provider: a service that manages and maintains a database of domain names and their correspoding IP addresses. 
* Domain Registrar: a company or organization accredited by domain name registries to allow individuals and businesses to register and manage domain names for their websites. For example, [Namecheap](https://www.namecheap.com/) or [Google Domains](https://domains.google/?gad=1&gclid=Cj0KCQjwzdOlBhCNARIsAPMwjbyNg3a3GcrQjbs3vBSDg4AWnnEkc96mJupyg8nqqiKyxVCyaePRXfAaAjRLEALw_wcB&gclsrc=aw.ds) are domain registrars. 
* Fermyon Domain Name: the default domain a Spin application gets allocated at creation time, which is a combindation of the application's name and a random string. For example, `finicky-whiskers-6cavavfh.fermyon.app`.

> * Subdomain: a hierarchical part of a larger domain name. It is added to the left of the primary domain name, separated by a dot (period). For example, `finicky-whiskers.fermyon.app` is a subdomain of `fermyon.app`

Reviewer: I think that perhaps while users are engaging in buying or setting up domains these terms may come in handy, I have placed them at the bottom and linked to them from within the page.

> * Subdomain: a hierarchical part of a larger domain name. It is added to the left of the primary domain name, separated by a dot (period). For example, `finicky-whiskers.fermyon.app` is a subdomain of `fermyon.app`

Reviewer: Also fixed up a few spelling and grammar points, I think this is worth keeping.

### Relationship Between Custom Domains And Spin Applications

> Currently there is a 1:1 relationship between a custom domain and a Spin application. This means you can apply a custom domain to at most one Spin application. Note that It can take up to 72 hours to delegate your domain. To move your custom domain between Spin applications, you must delete the custom domain and re-delegate it to Fermyon Cloud with the appropriate Spin application. We also do not support adding custom records at this time, aside from  Name Server (NS) records we automatically generate and a text (TXT) record if you have registered the domain with us previously. 

Reviewer 2: A "NS record we automatically generate" can't be a "custom record" can it?  Maybe I don't understand what "We also do not support" means in this context - are you saying these are forbidden and will cause Fermyon Custom Domains to stop working?

Reviewer: I am also not sure about this specific paragraph. @macolso can we get some more information about this please?

Author: I was trying to communicate that the user cannot add multiple records at this time. We allow the user to add a single record: the custom domain. So if they wanted to add both `www.coffee.com` and `coffee.com` to their Spin application, that wouldn't be possible. 

After the user provides their desired custom domain, we'll generate 4 nameserver records that the user must share with their domain registrar to delegate their domain to Fermyon cloud.

> Currently there is a 1:1 relationship between a custom domain and a Spin application. This means you can apply a custom domain to at most one Spin application. Note that It can take up to 72 hours to delegate your domain. To move your custom domain between Spin applications, you must delete the custom domain and re-delegate it to Fermyon Cloud with the appropriate Spin application. We also do not support adding custom records at this time, aside from  Name Server (NS) records we automatically generate and a text (TXT) record if you have registered the domain with us previously. 

Reviewer: Fixed in latest push.
